8329862: libjli GetApplicationHome cleanups and enhance jli tracing

MBaesken opened the pull request:
We have already good JLI tracing capabilities. But GetApplicationHome and GetApplicationHomeFromDll lack some tracing and should be enhanced.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8329862](https://bugs.openjdk.org/browse/JDK-8329862): libjli GetApplicationHome cleanups and enhance jli tracing (**Bug** - P4)


### Reviewers
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [4d998244](https://git.openjdk.org/jdk/pull/18699/files/4d99824469cb6206aa1a06888d0d7361dc6c1c0e)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18699/head:pull/18699` \
`$ git checkout pull/18699`

Update a local copy of the PR: \
`$ git checkout pull/18699` \
`$ git pull https://git.openjdk.org/jdk.git pull/18699/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18699`

View PR using the GUI difftool: \
`$ git pr show -t 18699`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18699.diff">https://git.openjdk.org/jdk/pull/18699.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18699#issuecomment-2045494980)

MBaesken commented:
Btw. I noticed that  GetModuleFileName return value is not checked, should this be done (at other locations in the JDk codebase it is done) ?
See  https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea

AlanBateman commented:
This looks very ad hoc. Not opposed to expanding the set of trace messages when this env variable is set but what is proposed here looks like trace messages that are left over from a debugging session. If we expand the tracing then I think it should be consistent with the existing tracing.

MBaesken commented:
>  If we expand the tracing then I think it should be consistent with the existing tracing.

I just added some tracing calls where I missed them. Do you think  I should do some more adjustments ?
Btw. is the  `GetApplicationHomeFromDll`   (which in fact is getting the image location from the jli shared lib location) approach for the launcher documented somewhere or is this something that  'just works' for a log time  ?

MBaesken commented:
> If we expand the tracing then I think it should be consistent with the existing tracing.

What exactly do you see as inconsistent ?

RealCLanger commented on src/java.base/windows/native/libjli/java_md.c:
```diff
@@ -321,6 +323,8 @@ GetJREPath(char *path, jint pathsize)
         }
     }
 
+    JLI_TraceLauncher("GetJREPath - attempt to get JRE location from shared lib of the image\n");
```
Maybe add a trace also in the USE_REGISTRY_LOOKUP section

AlanBateman replied:
> I am not sure if this even works any more. Maybe Alan could comment on this ?

The GetPublicJREHome function was removed at some point, I think JDK 9, as it didn't make sense to have in the OpenJDK project. However, Oracle installer did depend on it and will require a bit of effort to see if removing it now will cause an issue now or not.

AlanBateman replied:
@MBaesken I checked into the building with -DUSE_REGISTRY_LOOKUP as that compiled in code that the Oracle installer needed when it copied java.exe to somewhere. This is indeed left over code and can be removed from java_md.c.

MBaesken replied:
Hi Alan, thanks for checking this !
I removed the USE_REGISTRY_LOOKUP usages.

RealCLanger reviewed (approved):
To me this looks useful, although maybe the overall JLI tracing could be revisited.

RealCLanger commented:
> > If we expand the tracing then I think it should be consistent with the existing tracing.
> 
> What exactly do you see as inconsistent ?

Maybe the output of the tracing should look similar to other traces done through `JLI_TraceLauncher`? E.g. not mention method names but just tell what the program is doing... ?

MBaesken commented:
> > What exactly do you see as inconsistent ?
> 
> Maybe the output of the tracing should look similar to other traces done through `JLI_TraceLauncher`? E.g. not mention method names but just tell what the program is doing... ?

Okay, makes sense !  I will change that to make the output similar to existing traces.

AlanBateman commented:
I think this is way too ad hoc and looks like lefts over from a debugging session. So I don't think it should be integrated without stepping back and thinking more about what this tracing option is intended for.

AlanBateman commented:
/reviewers 2

MBaesken commented:
I adjusted the trace messages a bit to make the coding more consistent to the existing Jli trace code,
I removed the print of the function name because this is usually avoided in the existing cases.
I removed the  'did not succeed' Jli trace messages in GetJREPath because in case of success we would getJli trace  output  ('JRE path is <this-path>')  so missing success output means it failed.

RealCLanger commented:
Generally this looks more consistent now. Let's see what @AlanBateman thinks... 😄

MBaesken commented on src/java.base/unix/native/libjli/java_md_common.c:
```diff
@@ -82,6 +82,8 @@ GetApplicationHome(char *buf, jint bufsize)
 {
     const char *execname = GetExecName();
     if (execname != NULL) {
+        JLI_TraceLauncher("Launcher executable path is %s\n", execname);
```
> This trace seems a bit unsymmetric to its Windows counterpart. Maybe it should be left out here, too, since there is tracing in GetJREPath.

I removed the JLI trace output for the launcher exe path.

MBaesken commented:
Hi, any additional comments / reviews ?
Thanks Matthias

AlanBateman commented:
> Hi, any additional comments / reviews ? Thanks Matthias

It still looks like left over trace messages from a debugging session, need to think about about what tracing make sense here.

magicus commented:
What is the use case for this tracing functionality? I recently was quite helped by it when debugging static java launching. And in that case, the more logging the better. But that sounds like a very special case. Is this something that end users are supposed to use to help solve problems?

MBaesken commented:
> But that sounds like a very special case.

Not sure if it is really such a special case.  Currently both modes (getting the image path from launcher binary path , and getting the image path from 'the dll' / GetApplicationHomeFromDll  (libjli.so on Linux)  exist and are to my knowledge supported.  At least one jtreg test fails when the GetApplicationHomeFromDll  does not work any more.
So I think it is natural that both modes are also supported/augmented by some tracing.
We could probably reduce the trace message(s) added to just one if this is desired .

AlanBateman commented:
> @AlanBateman If this is not acceptable to add, I must once again iterate my question: What is the use case for the tracing functionality then?

The tracing was mostly for the options and timing when it was originally added.  I don't object to expanding the tracing. The issue is that the proposed additions are inconsistent with the existing trace messages, e.g.  GetJREPath's existing trace messages use "JRE path", the proposed messages switch to "JRE location".

BTW:  the vestiges of the "JRE" should really be removed. The GetJREPath function has no business looking for jre/lib as that location does not exist since JDK 9.

MBaesken commented:
I adjusted the output ot 'JRE path' .

>  The GetJREPath function has no business looking for jre/lib as that location does not exist since JDK 9.

Is the at/below

`/* Does the app ship a private JRE in <apphome>/jre directory? */`

part meant? This looks indeed obsolete .

tstuefe reviewed (approved):
LGTM

AlanBateman commented:
> 
> `/* Does the app ship a private JRE in <apphome>/jre directory? */`
> 
> part meant? This looks indeed obsolete .

Yes, this is is code that doesn't make sense since JDK 9 and should be removed/cleanup at some point. I suspect we had to leave it at the time because of the issue of installers copying java.exe into a shared location and expecting it to work with any JRE or JDK release, something that doesn't make sense now of course.

MBaesken commented:
I removed the mentioned 'private JRE'  check and also the related size check.

RealCLanger commented:
Still looks good. We should maybe change the synopsis (title) of the bug to something like "libjli GetApplicationHome cleanups"?

tstuefe commented:
Change still good

> Still looks good. We should maybe change the synopsis (title) of the bug to something like "libjli GetApplicationHome cleanups"?

Well, it does enhance tracing, so the title is not lying :)

MBaesken commented:
> Still looks good. We should maybe change the synopsis (title) of the bug to something like "libjli GetApplicationHome cleanups"?

I changed the synopsis because it was indeed now not really reflecting any more what the change does.

MBaesken commented:
Thanks for the reviews !

/integrate

AlanBateman commented:
> I removed the mentioned 'private JRE' check and also the related size check.

Good, I'll look at it as soon as I can. I suspect we'll need some follow on issues as there are several issues here.
